source/auth/auth.md

```diff
@@ -0,0 +1,2004 @@
+
+#### Errors
+
+Drivers SHOULD raise an error as early as possible when detecting invalid values in a credential. For instance, if a
+`mechanism_property` is specified for [MONGODB-CR](#mongodb-cr), the driver should raise an error indicating that the
+property does not apply.
+
+Drivers MUST raise an error if any required information for a mechanism is missing. For instance, if a `username` is not
+specified for SCRAM-SHA-256, the driver must raise an error indicating the the property is missing.
+
+#### Naming
+
+Naming of this information MUST be idiomatic to the driver's language/framework but still remain consistent. For
+instance, python would use "mechanism_properties" and .NET would use "MechanismProperties".
+
+Naming of mechanism properties MUST be case-insensitive. For instance, SERVICE_NAME and service_name refer to the same
+property.
+
+### Authentication
+
+A MongoClient instance MUST be considered a single logical connection to the server/deployment.
+
+Socket connections from a MongoClient to deployment members can be one of two types:
+
+- Monitoring-only socket: multi-threaded drivers maintain monitoring sockets separate from sockets in connection pools.
+- General-use socket: for multi-threaded drivers, these are sockets in connection pools used for (non-monitoring) user
+  operations; in single-threaded drivers, these are used for both monitoring and user operations.
+
+Authentication (including mechanism negotiation) MUST NOT happen on monitoring-only sockets.
+
+If one or more credentials are provided to a MongoClient, then whenever a general-use socket is opened, drivers MUST
+immediately conduct an authentication handshake over that socket.
+
+Drivers SHOULD require all credentials to be specified upon construction of the MongoClient. This is defined as eager
+authentication and drivers MUST support this mode.
+
+#### Authentication Handshake
+
+An authentication handshake consists of an initial `hello` or legacy hello command possibly followed by one or more
+authentication conversations.
+
+Drivers MUST follow the following steps for an authentication handshake:
+
+1. Upon opening a general-use socket to a server for a given MongoClient, drivers MUST issue a
+   [MongoDB Handshake](../mongodb-handshake/handshake.rst) immediately. This allows a driver to determine the server
+   type. If the `hello` or legacy hello of the MongoDB Handshake fails with an error, drivers MUST treat this as an
+   authentication error.
+1. If the server is of type RSArbiter, no authentication is possible and the handshake is complete.
+1. Inspect the value of `maxWireVersion`. If the value is greater than or equal to `6`, then the driver MUST use
+   `OP_MSG` for authentication. Otherwise, it MUST use `OP_QUERY`.
+1. If credentials exist: 3.1. A driver MUST authenticate with all credentials provided to the MongoClient. 3.2. A single
+   invalid credential is the same as all credentials being invalid.
```

Comment:
Numbering.

```suggestion
1. Upon opening a general-use socket to a server for a given MongoClient, drivers MUST issue a
   [MongoDB Handshake](../mongodb-handshake/handshake.rst) immediately. This allows a driver to determine the server
   type. If the `hello` or legacy hello of the MongoDB Handshake fails with an error, drivers MUST treat this as an
   authentication error.
2. If the server is of type RSArbiter, no authentication is possible and the handshake is complete.
3. Inspect the value of `maxWireVersion`. If the value is greater than or equal to `6`, then the driver MUST use
   `OP_MSG` for authentication. Otherwise, it MUST use `OP_QUERY`.
4. If credentials exist:
    1. A driver MUST authenticate with all credentials provided to the MongoClient.
    2. A single invalid credential is the same as all credentials being invalid.
```